Inside a switch statement, newline-before-return conflicts with lines-around-comment

<!--
This template is for bug reports. If you are reporting a bug, please continue on. If you are here for another reason, please see below:

1. To propose a new rule: http://eslint.org/docs/developer-guide/contributing/new-rules
2. To request a change: http://eslint.org/docs/developer-guide/contributing/changes
3. If you have any questions, please stop by our chatroom: https://gitter.im/eslint/eslint

Note that leaving sections blank will make it difficult for us to troubleshoot and we may have to close the issue.
-->

#### Background info
`lines-around-comment: ["error", {beforeLineComment: true, allowBlockStart: true}]` requires a blank line before any single-line comments, *except* where the comment is the first line of the block:

```
{
  // This comment is OK (not preceded by a newline, but first in block)

  // This comment is also OK (preceded by a newline)
  let a = 1;
  // This comment is NOT OK (not preceded by a newline)
}
```

`newline-before-return: "error"` requires a blank line before any return statements, *except* where the return is the only statement in a statement group.

```
function foo() {
  return 1;   // OK, return is the only statement
}

function bar() {
  const a = 1;

  return a;  // OK, return is preceded by a newline
}

function baz() {
  const a = 1;
  return a;   // NOT OK, return is not preceded by a new line
}
```

#### Problem
Inside a `SwitchCase`, there doesn't seem to be a way to get both of these rules to play nicely with each other.  The problem appears to be that `SwitchCase` is considered by `newline-before-return` to be a "statement group", yet it is not a block.

Because `SwitchCase` is not a block, `allowBlockStart` doesn't come into play, so `lines-around-comment` will expect a newline before all comments (including the first comment in each `SwitchCase`).

However `newline-before-return` will complain about this newline before a comment if the only statement in the `SwitchCase` is a return.

See examples below.

**What version of ESLint are you using?**
2.4.0

**What parser (default, Babel-ESLint, etc.) are you using?**
babel-eslint

**Please show your full configuration:**

```
/* eslint
     newline-before-return: 2
     lines-around-comment: [2, {beforeLineComment: true, allowBlockStart: true}]
*/
```

**What did you do? Please include the actual source code causing the issue.**
(for each of these examples, refer to the corresponding output below.  Only code in the default `SwitchCase` differs between each example.)
#### Example 1
```
function test(value) {
	switch (value) {
		case "foo":

			// do something (newline above required by lines-around-comment)
			break;

		case "bar":

			// do something else (newline above required by lines-around-comment)
			break;

		default:

			// bail out (newline above required by lines-around-comment)
			return -1;
	}

	// do some more stuff (newline above required by lines-around-comment)
	return 1;
}

test("foo");
```

#### Example 2
```
function test(value) {
	switch (value) {
		case "foo":

			// do something (newline above required by lines-around-comment)
			break;

		case "bar":

			// do something else (newline above required by lines-around-comment)
			break;

		default:
			// bail out (newline above prevented by newline-before-return)
			return -1;
	}

	// do some more stuff (newline above required by lines-around-comment)
	return 1;
}

test("foo");
```

#### Example 3
```
function test(value) {
	switch (value) {
		case "foo":

			// do something (newline above required by lines-around-comment)
			break;

		case "bar":

			// do something else (newline above required by lines-around-comment)
			break;

		default:
			return -1;
	}

	// do some more stuff (newline above required by lines-around-comment)
	return 1;
}

test("foo");
```
**What actually happened? Please include the actual, raw output from ESLint.**
#### Example 1
In the first example, `newline-before-return` complains about an unexpected newline in the default `SwitchCase`. This is because the return is the only statement in the group:

```
$ eslint foo.js

foo.js
  20:4  error  Unexpected newline before return statement  newline-before-return

✖ 1 problem (1 error, 0 warnings)
```

#### Example 2
In the second example, taking out the unexpected newline fixes the `newline-before-return` error; but triggers a `lines-around-comment` error instead.  This is because a `SwitchCase` is not a block, so `allowBlockStart` doesn't come into play:

```
$ eslint foo.js

/Users/scotto/GitHub/loot/foo.js
  18:4  error  Expected line before comment  lines-around-comment

✖ 1 problem (1 error, 0 warnings)
```

#### Example 3
The only way to get it to work is to remove the comment entirely. With no comment at all in the `SwitchCase`, the code passes.

**What did you expect to happen?**
I think the `allowBlockStart: true` option of `lines-around-comment` should also work for `SwitchCase`  OR there should be a new options such as `allowSwitchCase(Start|End)` to support this.

This would allow the above code to be written as:

```
/* eslint
     newline-before-return: 2
     lines-around-comment: [2, {beforeLineComment: true, allowSwitchCaseStart: true}]
*/
function test(value) {
	switch (value) {
		case "foo":
			// do something
			break;

		case "bar":
			// do something else
			break;

		default:
			// bail out
			return -1;
	}

	// do some more stuff (newline above required by lines-around-comment)
	return 1;
}

test("foo");
```

--- a/lib/rules/lines-around-comment.js
+++ b/lib/rules/lines-around-comment.js
@@ -165,7 +165,7 @@ module.exports = function(context) {
      * @returns {boolean} True if the comment is at block start.
      */
     function isCommentAtBlockStart(node) {
-        return isCommentAtParentStart(node, "ClassBody") || isCommentAtParentStart(node, "BlockStatement");
+        return isCommentAtParentStart(node, "ClassBody") || isCommentAtParentStart(node, "BlockStatement") || isCommentAtParentStart(node, "SwitchCase");
     }
 
     /**
@@ -174,7 +174,7 @@ module.exports = function(context) {
      * @returns {boolean} True if the comment is at block end.
      */
     function isCommentAtBlockEnd(node) {
-        return isCommentAtParentEnd(node, "ClassBody") || isCommentAtParentEnd(node, "BlockStatement");
+        return isCommentAtParentEnd(node, "ClassBody") || isCommentAtParentEnd(node, "BlockStatement") || isCommentAtParentEnd(node, "SwitchCase") || isCommentAtParentEnd(node, "SwitchStatement");
     }
 
     /**

--- a/tests/lib/rules/lines-around-comment.js
+++ b/tests/lib/rules/lines-around-comment.js
@@ -158,6 +158,34 @@ ruleTester.run("lines-around-comment", rule, {
             }]
         },
         {
+            code: "switch ('foo'){\ncase 'foo':\n// line at switch case start\nbreak;\n}",
+            options: [{
+                beforeLineComment: true,
+                allowBlockStart: true
+            }]
+        },
+        {
+            code: "switch ('foo'){\ncase 'foo':\n\n// line at switch case start\nbreak;\n}",
+            options: [{
+                beforeLineComment: true,
+                allowBlockStart: true
+            }]
+        },
+        {
+            code: "switch ('foo'){\ncase 'foo':\nbreak;\n\ndefault:\n// line at switch case start\nbreak;\n}",
+            options: [{
+                beforeLineComment: true,
+                allowBlockStart: true
+            }]
+        },
+        {
+            code: "switch ('foo'){\ncase 'foo':\nbreak;\n\ndefault:\n\n// line at switch case start\nbreak;\n}",
+            options: [{
+                beforeLineComment: true,
+                allowBlockStart: true
+            }]
+        },
+        {
             code: "function foo(){   \n/* block comment at block start */\nvar g = 1;\n}",
             options: [{
                 allowBlockStart: true
@@ -203,6 +231,30 @@ ruleTester.run("lines-around-comment", rule, {
             options: [{ allowBlockStart: true }],
             parserOptions: { ecmaVersion: 6 }
         },
+        {
+            code: "switch ('foo'){\ncase 'foo':\n/* block comment at switch case start */\nbreak;\n}",
+            options: [{
+                allowBlockStart: true
+            }]
+        },
+        {
+            code: "switch ('foo'){\ncase 'foo':\n\n/* block comment at switch case start */\nbreak;\n}",
+            options: [{
+                allowBlockStart: true
+            }]
+        },
+        {
+            code: "switch ('foo'){\ncase 'foo':\nbreak;\n\ndefault:\n/* block comment at switch case start */\nbreak;\n}",
+            options: [{
+                allowBlockStart: true
+            }]
+        },
+        {
+            code: "switch ('foo'){\ncase 'foo':\nbreak;\n\ndefault:\n\n/* block comment at switch case start */\nbreak;\n}",
+            options: [{
+                allowBlockStart: true
+            }]
+        },
 
         // check for block end comments
         {
@@ -242,6 +294,34 @@ ruleTester.run("lines-around-comment", rule, {
         },
         {
             code: "if(true){\nvar g = 1;\n\n// line at block end\n}",
+            options: [{
+                afterLineComment: true,
+                allowBlockEnd: true
+            }]
+        },
+        {
+            code: "switch ('foo'){\ncase 'foo':\nvar g = 1;\n\n// line at switch case end\n}",
+            options: [{
+                afterLineComment: true,
+                allowBlockEnd: true
+            }]
+        },
+        {
+            code: "switch ('foo'){\ncase 'foo':\nvar g = 1;\n\n// line at switch case end\n\n}",
+            options: [{
+                afterLineComment: true,
+                allowBlockEnd: true
+            }]
+        },
+        {
+            code: "switch ('foo'){\ncase 'foo':\nbreak;\n\ndefault:\nvar g = 1;\n\n// line at switch case end\n}",
+            options: [{
+                afterLineComment: true,
+                allowBlockEnd: true
+            }]
+        },
+        {
+            code: "switch ('foo'){\ncase 'foo':\nbreak;\n\ndefault:\nvar g = 1;\n\n// line at switch case end\n\n}",
             options: [{
                 afterLineComment: true,
                 allowBlockEnd: true
@@ -339,6 +419,34 @@ ruleTester.run("lines-around-comment", rule, {
                 allowBlockEnd: true
             }],
             parserOptions: { ecmaVersion: 6 }
+        },
+        {
+            code: "switch ('foo'){\ncase 'foo':\nvar g = 1;\n\n/* block comment at switch case end */\n}",
+            options: [{
+                afterBlockComment: true,
+                allowBlockEnd: true
+            }]
+        },
+        {
+            code: "switch ('foo'){\ncase 'foo':\nvar g = 1;\n\n/* block comment at switch case end */\n\n}",
+            options: [{
+                afterBlockComment: true,
+                allowBlockEnd: true
+            }]
+        },
+        {
+            code: "switch ('foo'){\ncase 'foo':\nbreak;\n\ndefault:\nvar g = 1;\n\n/* block comment at switch case end */\n}",
+            options: [{
+                afterBlockComment: true,
+                allowBlockEnd: true
+            }]
+        },
+        {
+            code: "switch ('foo'){\ncase 'foo':\nbreak;\n\ndefault:\nvar g = 1;\n\n/* block comment at switch case end */\n\n}",
+            options: [{
+                afterBlockComment: true,
+                allowBlockEnd: true
+            }]
         },
 
         // check for object start comments
@@ -777,6 +885,20 @@ ruleTester.run("lines-around-comment", rule, {
             errors: [{ message: afterMessage, type: "Line", line: 4 }]
         },
         {
+            code: "switch ('foo'){\ncase 'foo':\n// line at switch case start\nbreak;\n}",
+            options: [{
+                beforeLineComment: true
+            }],
+            errors: [{ message: beforeMessage, type: "Line", line: 3 }]
+        },
+        {
+            code: "switch ('foo'){\ncase 'foo':\nbreak;\n\ndefault:\n// line at switch case start\nbreak;\n}",
+            options: [{
+                beforeLineComment: true
+            }],
+            errors: [{ message: beforeMessage, type: "Line", line: 6 }]
+        },
+        {
             code: "while(true){\n// line at block start and end\n}",
             options: [{
                 afterLineComment: true,
@@ -791,6 +913,20 @@ ruleTester.run("lines-around-comment", rule, {
                 allowBlockEnd: true
             }],
             errors: [{ message: beforeMessage, type: "Line", line: 2 }]
+        },
+        {
+            code: "switch ('foo'){\ncase 'foo':\nvar g = 1;\n\n// line at switch case end\n}",
+            options: [{
+                afterLineComment: true
+            }],
+            errors: [{ message: afterMessage, type: "Line", line: 5 }]
+        },
+        {
+            code: "switch ('foo'){\ncase 'foo':\nbreak;\n\ndefault:\nvar g = 1;\n\n// line at switch case end\n}",
+            options: [{
+                afterLineComment: true
+            }],
+            errors: [{ message: afterMessage, type: "Line", line: 8 }]
         },
 
         // object start comments

